### ISSUE TITLE
Bluetooth Profile issues

### ISSUE THREAD

**Issue by efraimbart** [Android-L-Porting-Team/Android-L-Mako#10]
What steps will reproduce the problem?
1.Connect a Bluetooth audio device (one that only works for calls not music)
2.Make a call
3.

What is the expected output? What do you see instead?
Phone should recognize it as Bluetooth headset, phone only recognizes it as a regular bluetooth, calls don't route through it.

Please use labels and text to provide additional information.

**Comment by rhishavphukan** [comment]
I am facing a problem. When the phone is connected to the car via bluetooth, I can stream music but can't connect phone calls. Music icon shows and phone network icon doesn't show on the sony car head unit. So I am not able to make handsfree calls.
Have disconnected and reconnected the device from the head unit many times. In the bluetooth menu, on media audio shows up. 
http://forum.xda-developers.com/picture.php?albumid=9729&pictureid=45590

**Comment by cred05** [comment]
I am also unable to connect to my car's bluetooth. I will get a pin request, and it will say that I am paired with my car, but it doesn't actually connect. The cell signal icon doesn't show on my car's screen either. This is to connect to make phone calls only, my car does not have A2DP.

**Comment by fooznugget** [comment]
I have the same problem with my Mazda 3.  I can pair with the bluetooth fine, but when I go into the actual settings for the connection on the phone, the only profile that shows us is for media audio.  Typically, it will have media audio and another one for the phone.  Similarly the Car only shows the icon for the media audio, and not the phone icon.  I've tried unpairing and repairing several times as well with no change.

**Comment by mattjanicki** [comment]
I'm basically having the same problem described by cred05 with my Audi A6.

**Comment by paulthvt** [comment]
Facing the same issue with my Clio 3.2.

**Comment by Sykopompos** [comment]
We are looking into it but it might be an upstream issue with the developer preview itself

**Comment by canandroid** [comment]
Can confirm same issue. I don't see a request for access to the phonebook when I pair either.

**Comment by jackbeasley** [comment]
Same issue. The bluetooth menu in settings only lists the Media audio profile and calls are not routed through the car.

**Comment by rhishavphukan** [comment]
Hope this problem is resolved soon as I don't want to revert to KITKAT.

**Comment by rhishavphukan** [comment]
Has anyone tried beta 3?
Please inform if the above issues are solved.

**Comment by jonas-koeritz** [comment]
Habe you tried unpairing and pairing again? My Parrot Asteroid Classic paired flawlessly and A2DP and Phone Audio is working.

**Comment by rhishavphukan** [comment]
Yes have tried several times. Did not work.  jonas-koeritz did it work the first time you connected?

**Comment by jonas-koeritz** [comment]
After unpairing my parrot was connected in first try. Asked for Phonebook Access and everything went fine.

**Comment by fooznugget** [comment]
I had this problem on Beta 2 and Beta 3.

**Comment by canandroid** [comment]
Beta 3 Preview 3 here, paired and unpaired a number of times with no luck. The car did briefly flash something on its screen that I couldn't see - maybe trying to get the phonebook? I'll try to grab a logcat this afternoon and see if it's being denied or something. Media works just fine though.

**Comment by canandroid** [comment]
I have the same issue with Preview 5. I grabbed a logcat during pairing: http://pastebin.com/PdvuWKRM

**Comment by Sykopompos** [comment]
I'm going to look into this tomorrow

**Comment by stx8** [comment]
Yep, I also have this exact issue with a Subaru (cannot pair for calls but can stream). However, what's more strange is that my bluetooth headset worked fine for calls AND streaming audio and so does my GPS unit (standalone). it seems to be - if this thread is evidence - closely related to stock car decks. Strange though that streaming music works fine!

**Comment by Sykopompos** [comment]
I have been looking and can't figure this out because it matches up with 4.4.4 , the way cars work is that it uses three profiles. handsfree, headset and a2dp and one isn't connecting

**Comment by jonas-koeritz** [comment]
My Parrot Asteroid Classic pairs perfectly fine but I started to look for
problems and finally found one. Everything works but Battery display and
Network strength. Battery is always shown as fully discharged and Signal
strength is shown as full power (can't say if this is just because my
signal is full power on my short way home). Where is the code in Android
that controls all this stuff? E.g. avrcp display of album, artist and title
(it's missing album art since ages...never seen an android phone do this.).
Am 11.07.2014 02:56 schrieb "Mike Finnegan" notifications@github.com:

**Comment by AndersBerken** [comment]
I got mine paired with the VW car. No signal strength show no battery show (both works with 4.4.4).
I also noticed that I can't see who is calling (works with 4.4.4 KitKat).
I also noticed that it flips out and stop functioning to make calls. Car just showing message "EMPTY!" when I try to make call.

**Comment by Sykopompos** [comment]
I think I got it figured out http://forum.xda-developers.com/showpost.php?p=54092367&postcount=2793

**Comment by rhishavphukan** [comment]
Didn't work. Tried N4 and N5 files but it gives an error saying Bluetooth share has stopped working.

**Comment by rhishavphukan** [comment]
These files weren't missing. As I pasted them in system/lib it asked to replace existing. Couldnt use and hence use root explorer and set permissions to rw- r-- r--

**Comment by thunder-62** [comment]
Did the libs cure the issue?

**Comment by Darex2094** [comment]
Confirming that the libs didn't do it -- Also was asked if I wanted to overwrite.

**Comment by Sykopompos** [comment]
Yeah. I had derped late at night but I'm thinking bout something else to try

**Comment by rhishavphukan** [comment]
Sykopompos thanks for trying to smash the bug. Looking forward to get it working. Keep up the good work bruh

**Comment by Sykopompos** [comment]
This one got me stumped. I have been all up and down and under looking under every rock and I have yet to figure this out. Gonna keep looking though

**Comment by Sykopompos** [comment]
alright guys, give this a try and let me know
http://forum.xda-developers.com/showpost.php?p=54111272&postcount=2989

**Comment by thunder-62** [comment]
tried it with my BMW... 
Advanced Audio Distribution Profile (A2DP)            ok
Hands-Free Profile (HFP)                                        not ok

tried it with my Jabra Stone...
Advanced Audio Distribution Profile (A2DP)            ok
Headset Profile (HSP)                                              ok 

so the issue is with the Hands-Free Profile

**Comment by Sykopompos** [comment]
Can you get a log for the hands free?

**Comment by thunder-62** [comment]
Have a 12k lines logcat.... Try to make sense out of it before uploading it

reading http://developer.android.com/guide/topics/connectivity/bluetooth.html#Profiles , I understand that Headset an Handsfree is implemented in the same Android profile....

And almost forgot that L had changes in BT because they now allow also BT LE Client behaviour....

It seems like this is the relevant section:
I/BluetoothA2dpServiceJni( 4718): connectA2dpNative: sBluetoothA2dpInterface: 0x9c30037c
I/BluetoothA2dpServiceJni( 4718): bta2dp_connection_state_callback
D/CachedBluetoothDevice( 4435): Command sent successfully:CONNECT Address:00:0E:9F:DC:7B:E1 Profile:A2DP
D/BluetoothAdapterProperties( 4718): CONNECTION_STATE_CHANGE: 00:0E:9F:DC:7B:E1: 0 -> 1
D/BluetoothAdapterService(693628208)( 4718): Get Bonded Devices being called
D/BluetoothAdapterProperties( 4718): getBondedDevices: length=3
I/sf_frame_dur(  190): [com.android.settings/com.android.settings.bluetooth.BluetoothPairingDialog,51,3,5,3,17,1,0]
D/CachedBluetoothDevice( 4435): onProfileStateChanged: profile A2DP newProfileState 1
I/BluetoothA2dpServiceJni( 4718): bta2dp_connection_state_callback
D/BluetoothAdapterService(693628208)( 4718): Get Bonded Devices being called
D/BluetoothAdapterProperties( 4718): getBondedDevices: length=3
D/CachedBluetoothDevice( 4435): onProfileStateChanged: profile A2DP newProfileState 2
D/BluetoothAdapterProperties( 4718): CONNECTION_STATE_CHANGE: 00:0E:9F:DC:7B:E1: 1 -> 2
D/BluetoothAdapterService(693628208)( 4718): Get Bonded Devices being called
D/BluetoothAdapterProperties( 4718): getBondedDevices: length=3
I/BluetoothAvrcpServiceJni( 4718): btavrcp_remote_features_callback
I/BluetoothAvrcpServiceJni( 4718): btavrcp_register_notification_callback
I/BluetoothAvrcpServiceJni( 4718): registerNotificationRspPlayStatusNative: sBluetoothAvrcpInterface: 0x9c3003a4
I/BluetoothAvrcpServiceJni( 4718): btavrcp_remote_features_callback
I/BluetoothAvrcpServiceJni( 4718): btavrcp_register_notification_callback
I/BluetoothAvrcpServiceJni( 4718): registerNotificationRspTrackChangeNative: sBluetoothAvrcpInterface: 0x9c3003a4
I/BluetoothAvrcpServiceJni( 4718): btavrcp_register_notification_callback
I/BluetoothAvrcpServiceJni( 4718): registerNotificationRspPlayPosNative: sBluetoothAvrcpInterface: 0x9c3003a4
I/BluetoothHeadsetServiceJni( 4718): connectHfpNative: sBluetoothHfpInterface: 0x9c300338
D/CachedBluetoothDevice( 4435): onProfileStateChanged: profile HEADSET newProfileState 1
I/BluetoothHeadsetServiceJni( 4718): connection_state_callback
I/LocalBluetoothProfileManager( 4435): Failed to connect HEADSET device
D/CachedBluetoothDevice( 4435): onProfileStateChanged: profile HEADSET newProfileState 0
D/BluetoothManagerService(  607): disable(): mBluetooth = android.bluetooth.IBluetooth$Stub$Proxy@214cc9fd mBinding = fa
lse
D/BluetoothManagerService(  607): Message: 2
D/BluetoothManagerService(  607): Sending off request.
D/BluetoothAdapterState( 4718): CURRENT_STATE=ON, MESSAGE = USER_TURN_OFF
D/BluetoothAdapterProperties( 4718): Setting state to 13
I/BluetoothAdapterState( 4718): Bluetooth adapter state changed: 12-> 13
D/BluetoothAdapterService( 4718): Broadcasting updateAdapterState() to 1 receivers.
D/BluetoothAdapterProperties( 4718): onBluetoothDisable()
I/BluetoothAdapterProperties( 4718): adapterPropertyChangedCallback with type:7 len:4
D/BluetoothAdapterProperties( 4718): Scan Mode:20
I/BluetoothAdapterState( 4718): Entering PendingCommandState State: isTurningOn()=false, isTurningOff()=true
D/BluetoothAdapterState( 4718): CURRENT_STATE=PENDING, MESSAGE = BEGIN_DISABLE, isTurningOn=false, isTurningOff=true
D/BluetoothManagerService(  607): Message: 60
D/BluetoothManagerService(  607): MESSAGE_BLUETOOTH_STATE_CHANGE: prevState = 12, newState=13
D/BluetoothManagerService(  607): Bluetooth State Change Intent: 12 -> 13
D/BluetoothMapService( 4718): onReceive
D/BluetoothMapService( 4718): STATE_TURNING_OFF
D/BluetoothMapService( 4718): MAP Service closeService in
I/BtOppRfcommListener( 4718): BluetoothSocket listen thread finished

**Comment by Sykopompos** [comment]
that is the same thing I am seeing with the random redial issues with bluetooth when I read the logcat.
I'm still working on trying to figure this out 
http://forum.xda-developers.com/showpost.php?p=54126081&postcount=3160

**Comment by thunder-62** [comment]
Test Results:

4.1  Bluetooth won't turn on
4.2  Buetooth configuration FC
4.3  Bluetooth turns on but pairing only A2DP

**Comment by jonzey231** [comment]
Here's a video of the issue that we are all having. Just wanted to do whatever I could to help this project along:

http://youtu.be/7SLuGuqxlP0

**Comment by rhishavphukan** [comment]
Anyone tried beta 4? BT issues solved?

**Comment by thunder-62** [comment]
Unfortunately still there in beta4

**Comment by thunder-62** [comment]
Tried whether a N7(2013) with the official "L" preview connects with handsfree profile.

IT DOES NOT!  (Which makes sense as it has no phone capabilities)

maybe it could help to compare the N7 and N5 images....

**Comment by spupuz** [comment]
+1

**Comment by Sniper435** [comment]
Just wanted to report that I'm also seeing bluetooth errors when trying to use my Plantronics headset.
I'm running a clean install of Beta 4 - when I connect the headset it immediately redials the last number called, headset buttons appear non-responsive however audio does work.

**Comment by thunder-62** [comment]
@Sniper436: your issue is most likely described in issue https://github.com/Android-L-Porting-Team/Android-L-Mako/issues/45

**Comment by Sniper435** [comment]
@thunder-62 looks like you might be right - unless they are really the same bug.

**Comment by gruncho2004** [comment]
I have the same problem with a handsfree parrot ck3100. But works fine with sony smartwatch2, sony sbh20...

**Comment by ajayclk** [comment]
Same issue... It can play audio from my n4 but no voice calls

**Comment by AndersBerken** [comment]
No change on this issue with 4.9.1 either.

**Comment by gruncho2004** [comment]
Issue persists on 4.9.99.5.1

**Comment by Sykopompos** [comment]
Sadly I know. Trying to fix the Bluetooth bugs has become a personal quest of mine and it seems to have turned into the Iliad

**Comment by AndersBerken** [comment]
Good luck Sykopompos! You can do it! (Btw, can also confirm that erlier reported issues seems to persist in current version as well.

**Comment by AndersBerken** [comment]
As for the release with all the 9s: issue remain. Cant make a call from the car.

**Comment by jonzey231** [comment]
Going to do a clean install of Beta 5 and then give it a shot.

**Comment by rhishavphukan** [comment]
Any luck with B5?

**Comment by jonzey231** [comment]
Nope it still won't work. 😢
